[Dist Profiling] Enable profiling for gloo send/recv

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #52031 [Dist Profiling] Enable dist profiling for DDP
* **#52004 [Dist Profiling] Enable profiling for gloo send/recv**
* #51822 [Dist Profiling] Support shape recording for profiling collectives in gloo

Enables profiling of p2p collectives for Gloo. Modified/added relevant unittests.

Differential Revision: [D26347164](https://our.internmc.facebook.com/intern/diff/D26347164/)

rohan-varma committed:
[Dist Profiling] Enable profiling for gloo send/recv

Enables profiling of p2p collectives for Gloo. Modified/added relevant unittests.

Differential Revision: [D26347164](https://our.internmc.facebook.com/intern/diff/D26347164/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[Dist Profiling] Enable profiling for gloo send/recv"


Enables profiling of p2p collectives for Gloo. Modified/added relevant unittests.

Differential Revision: [D26347164](https://our.internmc.facebook.com/intern/diff/D26347164/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[Dist Profiling] Enable profiling for gloo send/recv"


Enables profiling of p2p collectives for Gloo. Modified/added relevant unittests.

Differential Revision: [D26347164](https://our.internmc.facebook.com/intern/diff/D26347164/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[Dist Profiling] Enable profiling for gloo send/recv"


Enables profiling of p2p collectives for Gloo. Modified/added relevant unittests.

Differential Revision: [D26347164](https://our.internmc.facebook.com/intern/diff/D26347164/)

[ghstack-poisoned]

mrzzd commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -412,8 +418,16 @@ void ProcessGroupGloo::SendWork::abort() {
+          -1,
+          OpType::UNKNOWN,
+          profilingTitle,
+          c10::optional<std::vector<at::Tensor>>({tensor})),
```
For receive op, the output tensor does not have it's final dimension until end of the operation right? Then does it record the operator shape correctly?

rohan-varma replied:
I think if the user does something like:
```
# on rank 1
dist.recv(tensor, src=0)
```
Then `tensor` is expected to be correctly sized with whatever tensor rank 0 is sending to 1. This appears to be true in all our tests in `distributed_test.py` and also the docs imply this: https://github.com/pytorch/pytorch/blob/master/torch/distributed/distributed_c10d.py#L819

mrzzd replied:
I could not understand it from the docs. If the API does not require that explicitly, could you double check that in case the output tensor is not sized prior to call, recv does not work properly?

rohan-varma replied:
Sure, I just re-verified this in the tests. In `test_send_recv`, if you change:
`output_tensor = _build_tensor(recv_size, value=-1)` to 
`output_tensor = _build_tensor(recv_size -1, value=-1)` we run into the following error:

```
W0211 16:35:58.875078 813884 ExceptionTracer.cpp:185] Invalid trace stack for exception of type: gl
oo::EnforceNotMet
terminate called after throwing an instance of 'gloo::EnforceNotMet'
  what():  [enforce fail at gloo/transport/tcp/pair.cc:510] op.preamble.length <= op.nbytes. 4 vs 0
/data/users/rvarm1/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/distributed_nccl_fork#b
inary,link-tree/torch/testing/_internal/common_distributed.py:292: Caught exception:
Traceback (most recent call last):
  File "/data/users/rvarm1/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/distributed_ncc
l_fork#binary,link-tree/torch/testing/_internal/common_distributed.py", line 289, in wrapper
    fn()
  File "/data/users/rvarm1/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/distributed_ncc
l_fork#binary,link-tree/torch/testing/_internal/distributed/distributed_test.py", line 885, in test
_send_recv
    dist.recv(output_tensor, src)
  File "/data/users/rvarm1/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/distributed_ncc
l_fork#binary,link-tree/torch/distributed/distributed_c10d.py", line 850, in recv
    pg.recv([tensor], src, tag).wait()
RuntimeError: [gloo/transport/tcp/pair.cc:589] Read error [2401:db00:21:6045:face:0:5:0]:12722: Con
nection reset by peer
exiting process with exit code: 10
Process 0 terminated with exit code 10, terminating remaining processes.
ERROR
```

However, if the `recv` tensor size is larger, then the recv succeeds, and we record this larger shape. However I believe that this is the desired behavior, since the larger tensor is what is input into `gloo:recv` buffer.

rohan-varma committed:
Update on "[Dist Profiling] Enable profiling for gloo send/recv"


Enables profiling of p2p collectives for Gloo. Modified/added relevant unittests.

Differential Revision: [D26347164](https://our.internmc.facebook.com/intern/diff/D26347164/)

[ghstack-poisoned]

mrzzd reviewed (approved):
Thanks Rohan!
